File: ios/Runner/FcmTokenRefreshEventApi.swift
//
//  FcmTokenRefreshUri.swift
//  Runner
//
//  Created by Kazuya Chikamatsu on 2023/08/14.
//
import Flutter

class FcmTokenRefreshEventApi: NSObject {
    let channelName = "net.chikach.submon.event/fcm_token_refresh"
    var channel: FlutterEventChannel
    var eventSink: FlutterEventSink?
    var pendingFcmToken: String?
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        channel = FlutterEventChannel(name: channelName, binaryMessenger: binaryMessenger)
    }
    
    func initHandler() {
        channel.setStreamHandler(self)
    }
    
    func onFcmTokenRefresh(token: String) {
        if (eventSink != nil) {
            eventSink!(token)
        } else {
            pendingFcmToken = token
        }
    }
}

extension FcmTokenRefreshEventApi: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        if (pendingFcmToken != nil) {
            eventSink!(pendingFcmToken)
            pendingFcmToken = nil
        }
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/net/chikach/submon/eventapi/FcmTokenRefreshEventApi.kt
package net.chikach.submon.eventapi

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class FcmTokenRefreshEventApi(binaryMessenger: BinaryMessenger) : EventChannel.StreamHandler {
    private val channel = EventChannel(binaryMessenger, CHANNEL_NAME)

    companion object {
        const val CHANNEL_NAME = "net.chikach.submon.event/fcm_token_refresh"
    }

    private var eventSink: EventChannel.EventSink? = null
    private var pendingData: String? = null

    fun initHandler() {
        channel.setStreamHandler(this)
    }

    fun onFcmTokenRefresh(token: String) {
        if (eventSink == null) {
            pendingData = token
        } else {
            eventSink!!.success(token)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        if (pendingData != null) {
            eventSink!!.success(pendingData)
            pendingData = null
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}